### PR TITLE
ampl-mp 4.0.0 ampl-asl 1.0.0 (new formula)

### PR DESCRIPTION
**Pull request by chenrui333**
Created with `brew bump-formula-pr`.

- [x] `resource` blocks have been checked for updates.
<details>
  <summary>release notes</summary>
  <pre># First tagged release since 2016

## Overview
* Automatic reformulation framework
  * Logical, combinatorial and conic constraints are reformulated into basic constructs suitable for a given solver, see https://mp.ampl.com/
  * Out-multiplication of quadratics, as typical for MIQP solvers
  * Piecewise-linear approximation of univariate nonlinear functions
* 9 MILP/MIQP solvers implemented in MP, they are now default AMPL binaries
* Multi-objective emulator #239
* Reformulation explorer #232
* Automatic solution check #200
* NL Writer library added in C/C++ and Python #30
* Expression tree output, suitable for (MI)NLP solver API, added in #137

## Pull requests
* Not all Linux have fpu_control.h (that's a glibc specific header) by @oswaldo4 in https://github.com/ampl/mp/pull/102
* Fixwindows by @mapgccv in https://github.com/ampl/mp/pull/117
* Allow skip building tests with -DBUILD_TESTING=off by @zhihaoy in https://github.com/ampl/mp/pull/123
* find_library(DL_LIBRARY dl) is wrong by @zhihaoy in https://github.com/ampl/mp/pull/124
* CMake builtin c++11 flags by @zhihaoy in https://github.com/ampl/mp/pull/125
* Update to aslinterface to integrate complementarity constraints by @tmigot in https://github.com/ampl/mp/pull/112
* 64 bit windows build  by @mapgccv in https://github.com/ampl/mp/pull/127
* Update fmt calls in the MySQL test by @jamesjer in https://github.com/ampl/mp/pull/128
* Update the gecode interface for gecode 5.x by @jamesjer in https://github.com/ampl/mp/pull/129
* avoid load of misaligned address warning in BinaryReader by @svigerske in https://github.com/ampl/mp/pull/134
* Added MIP backend by @mapgccv in https://github.com/ampl/mp/pull/136
* MIPBackend: more changes by @glebbelov in https://github.com/ampl/mp/pull/137
* Avoid jacop-related crashes due to destructor execution order by @jamesjer in https://github.com/ampl/mp/pull/133
* Feature/options by @mapgccv in https://github.com/ampl/mp/pull/140
* make namespace explicit when accessing fmt from fmt::internal by @svigerske in https://github.com/ampl/mp/pull/175
* Ortools by @mapgccv in https://github.com/ampl/mp/pull/176
* Revert "Ortools" by @mapgccv in https://github.com/ampl/mp/pull/177
* Merge preliminary ortools driver by @mapgccv in https://github.com/ampl/mp/pull/178
* Revert "Ortools" by @glebbelov in https://github.com/ampl/mp/pull/179
* Revert "Revert "Ortools"" Last of 5 by @glebbelov in https://github.com/ampl/mp/pull/181
* Revert "Revert "Ortools"" by @glebbelov in https://github.com/ampl/mp/pull/182
* Revert "Ortools": -5 of 5 by @glebbelov in https://github.com/ampl/mp/pull/183
* Update model-guide.rst by @4er4er4er in https://github.com/ampl/mp/pull/186
* Mosek driver new features by @szelidvihar in https://github.com/ampl/mp/pull/190
* Draft: Add SCIP driver by @jurgen-lentz in https://github.com/ampl/mp/pull/205
* change scip to scipmp by @jurgen-lentz in https://github.com/ampl/mp/pull/206
* Draft: GCG driver by @jurgen-lentz in https://github.com/ampl/mp/pull/207
* GCG driver by @jurgen-lentz in https://github.com/ampl/mp/pull/212
* More migration to <chrono> by @jamesjer in https://github.com/ampl/mp/pull/235

## New Contributors
* @oswaldo4 made their first contribution in https://github.com/ampl/mp/pull/102
* @mapgccv made their first contribution in https://github.com/ampl/mp/pull/117
* @zhihaoy made their first contribution in https://github.com/ampl/mp/pull/123
* @tmigot made their first contribution in https://github.com/ampl/mp/pull/112
* @jamesjer made their first contribution in https://github.com/ampl/mp/pull/128
* @svigerske made their first contribution in https://github.com/ampl/mp/pull/134
* @glebbelov made their first contribution in https://github.com/ampl/mp/pull/137
* @4er4er4er made their first contribution in https://github.com/ampl/mp/pull/186
* @szelidvihar made their first contribution in https://github.com/ampl/mp/pull/190
* @jurgen-lentz made their first contribution in https://github.com/ampl/mp/pull/205

**Full Changelog**: https://github.com/ampl/mp/compare/3.1.0...v4.0.0</pre>
</details>

### CODE REVIEW

**Comment by chenrui333** [comment]
```
  [ 95%] Building CXX object test/CMakeFiles/converter-mip-test.dir/converter-mip-test.cpp.o
  cd /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/build/test && /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/shims/linux/super/g++-11 -DFMT_USE_FILE_DESCRIPTORS -DMP_DATE=20240724 -DMP_SYSINFO="\"Linux x86_64\"" -DMP_TEST_DATA_DIR=\"/tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/test/data\" -DMP_USE_ATOMIC -DMP_USE_HASH -DMP_USE_UNIQUE_PTR -I/tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/include -I/tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/test/../thirdparty -I/tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/src -Wall -Wextra -pedantic -O3 -DNDEBUG -std=c++17 -fPIE -fno-delete-null-pointer-checks -MD -MT test/CMakeFiles/converter-mip-test.dir/converter-mip-test.cpp.o -MF CMakeFiles/converter-mip-test.dir/converter-mip-test.cpp.o.d -o CMakeFiles/converter-mip-test.dir/converter-mip-test.cpp.o -c /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/test/converter-mip-test.cpp
  In file included from /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/include/mp/flat/converter.h:22,
                   from /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/test/converter-flat-test.h:10,
                   from /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/test/converter-flat-test.cpp:4:
  /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/include/mp/flat/constr_2_expr.h:400:13: error: explicit specialization in non-namespace scope ‘class mp::Constraints2Expr<Impl>’
    400 |   template <>
        |             ^
  In file included from /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/test/converter-flat-test.h:8,
                   from /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/test/converter-flat-test.cpp:4:
  /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/test/converter-flat-test.h: In static member function ‘static mp::ConstraintAcceptanceLevel TestBackendAcceptingConstraints<Constraint>::AcceptanceLevel(const Constraint*)’:
  /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/test/converter-flat-test.h:38:33: error: ‘mp::ConstraintAcceptanceLevel::mp’ has not been declared
     38 |   ACCEPT_CONSTRAINT(Constraint, mp::Recommended, mp::CG_Default)
        |                                 ^~
  /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/include/mp/flat/model_api_base.h:91:41: note: in definition of macro ‘ACCEPT_CONSTRAINT’
     91 | { return mp::ConstraintAcceptanceLevel::level; } \
        |                                         ^~~~~
  /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/test/converter-flat-test.h: In static member function ‘static mp::ConstraintAcceptanceLevel TestBackendAcceptingConstraints<Constraint>::AcceptanceLevel(const LinConEQ*)’:
  /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/test/converter-flat-test.h:39:35: error: ‘mp::ConstraintAcceptanceLevel::mp’ has not been declared
     39 |   ACCEPT_CONSTRAINT(mp::LinConEQ, mp::Recommended, mp::CG_Default)
        |                                   ^~
  /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/include/mp/flat/model_api_base.h:91:41: note: in definition of macro ‘ACCEPT_CONSTRAINT’
     91 | { return mp::ConstraintAcceptanceLevel::level; } \
        |                                         ^~~~~
  In file included from /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/include/mp/flat/converter.h:22,
                   from /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/include/mp/flat/redef/MIP/converter_mip.h:6,
                   from /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/test/converter-mip-test.h:7,
                   from /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/test/converter-mip-test.cpp:25:
  /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/include/mp/flat/constr_2_expr.h:400:13: error: explicit specialization in non-namespace scope ‘class mp::Constraints2Expr<Impl>’
    400 |   template <>
        |             ^
  In file included from /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/include/mp/flat/item_keeper.h:7,
                   from /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/include/mp/flat/constr_keeper.h:15,
                   from /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/include/mp/flat/converter_model.h:14,
                   from /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/include/mp/flat/converter.h:14,
                   from /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/include/mp/flat/redef/MIP/converter_mip.h:6,
                   from /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/test/converter-mip-test.h:7,
                   from /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/test/converter-mip-test.cpp:25:
  /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/test/converter-mip-test.h: In static member function ‘static mp::ConstraintAcceptanceLevel mip_converter_test::MIPInstanceBackend::AcceptanceLevel(const LinConEQ*)’:
  /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/test/converter-mip-test.h:127:35: error: ‘mp::ConstraintAcceptanceLevel::mp’ has not been declared
    127 |   ACCEPT_CONSTRAINT(mp::LinConEQ, mp::Recommended, mp::CG_Default)
        |                                   ^~
  /tmp/ampl-mp-20241108-3764-oi4a54/mp-4.0.0/include/mp/flat/model_api_base.h:91:41: note: in definition of macro ‘ACCEPT_CONSTRAINT’
     91 | { return mp::ConstraintAcceptanceLevel::level; } \
        |                                         ^~~~~
```

**Comment by chenrui333** [comment]
asl got removed from mp, https://github.com/ampl/asl/issues/21

File: Formula/i/ipopt.rb [hunk 1]
```diff
@@ -18,7 +19,7 @@ class Ipopt < Formula
 
   depends_on "openjdk" => :build
   depends_on "pkg-config" => [:build, :test]
-  depends_on "ampl-mp"
+  depends_on "ampl-mp@3"
```
Comment (SMillerDev) [inline Formula/i/ipopt.rb:22]: When can we switch back? Is there an issue upstream to update?

Comment (chenrui333) [reply Formula/i/ipopt.rb:22]: asl got removed in mp 4, so have to use mp3 for now, https://github.com/coin-or/Ipopt/issues/802

**Review by MikeMcQuaid — APPROVED** [review]

File: Formula/i/ipopt.rb (hunk 1 above)
Comment (cho-m) [reply Formula/i/ipopt.rb:22]: Does https://github.com/ampl/asl/releases/tag/v1.0.0 work instead?

File: Formula/a/ampl-mp.rb [hunk 2]
```diff
@@ -23,49 +22,33 @@ class AmplMp < Formula
...
-
-  # Removes Google Benchmark - as already done so upstream
-  # All it did was conflict with the google-benchmark formula
-  patch do
-    url "https://github.com/ampl/mp/commit/96e332bb8cb7ba925e3ac947d6df515496027eed.patch?full_index=1"
-    sha256 "1a4ef4cd1f4e8b959c20518f8f00994ef577e74e05824b2d1b241b1c3c1f84eb"
-  end
-
-  # Install missing header files, remove in > 3.1.0
-  # https://github.com/ampl/mp/issues/110
```
Comment (stefanb) [inline Formula/a/ampl-mp.rb:39]: Descriptions of all 3 pathches were not copied over to `ampl-mp@3.rb`. 
Was it intentional?
IMO it is not ok to have patches without some description.

File: Formula/a/ampl-mp@3.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    url "https://github.com/dpo/miniampl/archive/refs/tags/v1.0.tar.gz"
+    sha256 "b836dbf1208426f4bd93d6d79d632c6f5619054279ac33453825e036a915c675"
+  end
+
+  patch do
+    url "https://github.com/ampl/mp/commit/96e332bb8cb7ba925e3ac947d6df515496027eed.patch?full_index=1"
+    sha256 "1a4ef4cd1f4e8b959c20518f8f00994ef577e74e05824b2d1b241b1c3c1f84eb"
+  end
+
+  patch do
```
Comment (stefanb) [inline Formula/a/ampl-mp@3.rb:27]:
```suggestion
  # Install missing header files, remove in > 3.1.0
  # https://github.com/ampl/mp/issues/110
  patch do
```

File: Formula/a/ampl-mp@3.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  keg_only :versioned_formula
+
+  depends_on "cmake" => :build
+
+  resource "miniampl" do
+    url "https://github.com/dpo/miniampl/archive/refs/tags/v1.0.tar.gz"
+    sha256 "b836dbf1208426f4bd93d6d79d632c6f5619054279ac33453825e036a915c675"
+  end
+
+  patch do
```
Comment (stefanb) [inline Formula/a/ampl-mp@3.rb:22]:
```suggestion
  # Removes Google Benchmark - as already done so upstream
  # All it did was conflict with the google-benchmark formula
  patch do
```

File: Formula/a/ampl-mp@3.rb [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    url "https://github.com/ampl/mp/commit/96e332bb8cb7ba925e3ac947d6df515496027eed.patch?full_index=1"
+    sha256 "1a4ef4cd1f4e8b959c20518f8f00994ef577e74e05824b2d1b241b1c3c1f84eb"
+  end
+
+  patch do
+    url "https://github.com/ampl/mp/commit/8183be3e486d38d281e0c5a02a1ea4239695035e.patch?full_index=1"
+    sha256 "6b37201f1d0d6dba591e7e1b81fb16d2694d118605c92c422dcdaaedb463c367"
+  end
+
+  patch do
```
Comment (stefanb) [inline Formula/a/ampl-mp@3.rb:32]:
```suggestion
  # Backport fmt header update. Remove in the next release
  # https://github.com/ampl/mp/issues/108
  patch do
```

File: Formula/i/ipopt.rb (hunk 1 above)
Comment (chenrui333) [reply Formula/i/ipopt.rb]: yeah, it works fine for ipopt, so swap to use ampl-asl instead.

File: Formula/a/ampl-mp@3.rb (hunk 3 above)
Comment (chenrui333) [reply Formula/a/ampl-mp@3.rb]: yeah, I forgot to copy the patch desc over, but now, I just dropped ampl-mp in favor of ampl-asl.

**Comment by chenrui333** [comment]
dropped ampl-mp@3 in favor of ampl-asl (it builds/tests out fine for ipopt)

File: Formula/a/ampl-mp.rb (hunk 2 above)
Comment (chenrui333) [reply Formula/a/ampl-mp.rb:39]: agree, I think I forgot to copied while tweaking the versioned formula for ipopt, anyway after having upstream released 1.0.0 for ampl-asl, we no longer need ampl-mp@3

**Comment by SMillerDev** [comment]
```
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/shims/linux/super/gcc-11 -DNo_Control87 -Dasl_EXPORTS -I/tmp/ampl-asl-20241122-3962-tpnem/asl-1.0.0/build/include -I/tmp/ampl-asl-20241122-3962-tpnem/asl-1.0.0/src/solvers -m64 -O3 -DNDEBUG -flto=auto -fno-fat-lto-objects -fPIC -Wno-unused-result -Wno-parentheses -MD -MT CMakeFiles/asl.dir/src/solvers/xp1known.c.o -MF CMakeFiles/asl.dir/src/solvers/xp1known.c.o.d -o CMakeFiles/asl.dir/src/solvers/xp1known.c.o -c /tmp/ampl-asl-20241122-3962-tpnem/asl-1.0.0/src/solvers/xp1known.c
[ 97%] Building C object CMakeFiles/asl.dir/src/solvers/xp2known.c.o
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/shims/linux/super/gcc-11 -DNo_Control87 -Dasl_EXPORTS -I/tmp/ampl-asl-20241122-3962-tpnem/asl-1.0.0/build/include -I/tmp/ampl-asl-20241122-3962-tpnem/asl-1.0.0/src/solvers -m64 -O3 -DNDEBUG -flto=auto -fno-fat-lto-objects -fPIC -Wno-unused-result -Wno-parentheses -MD -MT CMakeFiles/asl.dir/src/solvers/xp2known.c.o -MF CMakeFiles/asl.dir/src/solvers/xp2known.c.o.d -o CMakeFiles/asl.dir/src/solvers/xp2known.c.o -c /tmp/ampl-asl-20241122-3962-tpnem/asl-1.0.0/src/solvers/xp2known.c
[ 98%] Linking C shared library bin/libasl.so
/home/linuxbrew/.linuxbrew/opt/cmake/bin/cmake -E cmake_link_script CMakeFiles/asl.dir/link.txt --verbose=1
./src/solvers/fgh_read.c:101:15: error: function ‘f2_HOL_ASL’ redeclared as variable
./src/solvers/rops2.c:1019:1: note: previously declared here
lto1: fatal error: errors during merging of translation units
compilation terminated.
lto-wrapper: fatal error: gcc-11 returned 1 exit status
compilation terminated.
/usr/bin/ld: error: lto-wrapper failed
collect2: error: ld returned 1 exit status
```

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by cho-m** [comment]
Could try disabling LTO on Linux

**Review by cho-m — APPROVED** [review]

**Pull request merged**